### PR TITLE
feat(dimensions): create composition function

### PR DESCRIPTION
**Pull request by nekosaur**
## Description

[**Notion Task**](https://www.notion.so/vuetify/E-Measureable-1fd56611f97f4c85ad6461d988eec512)

measurable mixin rewritten as a factory that generates a composition function

## TODO
- [x] I don't like the word measurement. It's long and maybe a bit diffuse? Feel free to come up with better ones.

## Motivation and Context

vue 3.0

## How Has This Been Tested?

playground, unit tests

## Markup:

<details>

```vue
<script lang="ts">
  import { mergeProps, h, defineComponent } from 'vue'
  import { measurementFactory } from '../src/effects/measurement'

  const { measurementProps, useMeasurement } = measurementFactory('width', 'height')

  export default defineComponent({
    props: {
      ...measurementProps({
        width: 100,
        height: 200,
      }),
    },
    setup (props) {
      const { measurement } = useMeasurement(props)

      return () => [
        h('div', mergeProps({
          style: {
            background: '#ff00ff',
          },
        }, measurement.value), ['foo']),
      ]
    },
  })
</script>
```
</details>

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Improvement/refactoring (non-breaking change that doesn't add any features but makes things better)

## Checklist:

- [x] The PR title is no longer than 64 characters.
- [x] The PR is submitted to the correct branch (`master` for bug fixes and documentation updates, `dev` for new features and backwards compatible changes and `next` for non-backwards compatible changes).
- [x] My code follows the code style of this project.
- [ ] I've added relevant changes to the documentation (applies to new features and breaking changes in core library)

### CODE REVIEW

File: packages/vuetify/src/effects/__tests__/measurement.spec.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+      expect(measurement.value.style.minWidth).toBeDefined()
+      expect(measurement.value.style.minHeight).toBeDefined()
+    })
+  })
+
+  describe('useMeasurement', () => {
+    it('should have proper styles', () => {
+      const { useMeasurement } = measurementFactory()
+
+      expect(useMeasurement(reactive({ width: 100 })).measurement.value).toEqual({
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/effects/__tests__/measurement.spec.ts]: No need for `reactive` here, but there could be reactivity test instead:

```
const x = reactive({ width: 100 })
const { measurement } = useMeasurement(x)

expect.measurement.value.toEqual(...)

x.width = null
expect.measurement.value.toEqual(...)
```

File: packages/vuetify/src/effects/__tests__/measurement.spec.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        style: {
+          width: '100px',
+          height: '200px',
+        },
+      })
+    })
+  })
+
+  describe('measurementProps', () => {
+    it('should have correct structure', () => {
```
Comment (jacekkarczmarczyk) [inline packages/vuetify/src/effects/__tests__/measurement.spec.ts]: maybe add test for all props `measurementFactory().measurementProps()`?

File: packages/vuetify/src/effects/__tests__/dimensions.spec.ts [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+      expect(dimensions.value.style.height).toBeDefined()
+      expect(dimensions.value.style.maxWidth).toBeDefined()
+      expect(dimensions.value.style.maxHeight).toBeDefined()
+      expect(dimensions.value.style.minWidth).toBeDefined()
+      expect(dimensions.value.style.minHeight).toBeDefined()
+    })
+  })
+
+  describe('useDimensions', () => {
+    it('should have proper styles', () => {
```
Comment (johnleider) [inline packages/vuetify/src/effects/__tests__/dimensions.spec.ts]: Could be simplified to a loop to reduce vertical space

File: packages/vuetify/src/effects/__tests__/dimensions.spec.ts [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        style: {
+          width: '100px',
+          height: '200px',
+        },
+      })
+    })
+
+    it('should be reactive', () => {
+      const { useDimensions } = dimensionsFactory()
+
```
Comment (johnleider) [inline packages/vuetify/src/effects/__tests__/dimensions.spec.ts]:
```suggestion
```

File: packages/vuetify/src/effects/__tests__/dimensions.spec.ts [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+      expect(useDimensions({ width: 100, height: '200' }).dimensions.value).toEqual({
+        style: {
+          width: '100px',
+          height: '200px',
+        },
+      })
+    })
+
+    it('should be reactive', () => {
```
Comment (johnleider) [inline packages/vuetify/src/effects/__tests__/dimensions.spec.ts]: Could be simplified to a loop to reduce vertical space

File: packages/vuetify/src/effects/dimensions.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
+import { computed, Prop } from 'vue'
+import { convertToUnit } from '../util/helpers'
+
+const allDimensionsProps = {
+  height: {
```
Comment (johnleider) [inline packages/vuetify/src/effects/dimensions.ts]: Should be alpha order

File: packages/vuetify/src/effects/__tests__/dimensions.spec.ts (hunk 3 above)
Comment (nekosaur) [reply packages/vuetify/src/effects/__tests__/dimensions.spec.ts]: Is this a suggestion only? I prefer to be explicit in tests.

Comment (KaelWD) [reply packages/vuetify/src/effects/__tests__/dimensions.spec.ts]: https://jestjs.io/docs/en/api#testeachtablename-fn-timeout

Comment (nekosaur) [reply packages/vuetify/src/effects/__tests__/dimensions.spec.ts]: nice, didn't know jest had support for test arguments

File: packages/vuetify/src/composables/__tests__/dimensions.spec.ts [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+      expect(dimensions.value.style.minWidth).toBeDefined()
+      expect(dimensions.value.style.minHeight).toBeDefined()
+    })
+  })
+
+  describe('useDimensions', () => {
+    it.each([
+      [{ width: 100 }, { width: '100px' }],
+      [{ width: 100, height: '' }, { width: '100px' }],
+      [{ width: 100, height: null }, { width: '100px' }],
```
Comment (KaelWD) [inline packages/vuetify/src/composables/__tests__/dimensions.spec.ts:47]: Type error, `PropValue` doesn't accept `null`

File: packages/vuetify/src/composables/__tests__/dimensions.spec.ts [hunk 8]
```diff
@@ -0,0 +1,147 @@
+import { reactive } from 'vue'
+
+import {
+  dimensionsFactory,
+} from '../dimensions'
+
+describe('dimensions.ts', () => {
```
Comment (KaelWD) [inline packages/vuetify/src/composables/__tests__/dimensions.spec.ts:7]: Don't really need the top group, jest groups everything by file anyway
image

**Review by KaelWD — APPROVED** [review]

**Pull request merged**